corehq/tests/test_classcleanup.py

```diff
@@ -0,0 +1,142 @@
+from functools import wraps
+from unittest import TestCase
+
+from nose.plugins import PluginTester
+from testil import eq
+
+from .noseplugins import classcleanup as mod
+
+
+class TestClassCleanupPlugin(PluginTester, TestCase):
+    activate = ''  # Activate option not needed. Plugin is always enabled.
+    plugins = [mod.ClassCleanupPlugin()]
+
+    def setUp(self):
+        pass  # super().setUp() is called by self.run_with_errors(...)
+
+    def makeSuite(self):
+        class Test(TestCase):
+            @classmethod
+            @maybe_error(self)
+            def setUpClass(cls):
+                cls.addClassCleanup(lambda: self.result.append("cleanup"))
+
+            @maybe_error(self)
+            def setUp(self):
+                pass
+
+            @maybe_error(self)
+            def runTest(self):
+                pass
+
+            @maybe_error(self)
+            def tearDown(self):
+                pass
+
+            @classmethod
+            @maybe_error(self)
+            def tearDownClass(cls):
+                pass
+
+        return [Test()]
+
+    def run_with_errors(self, *errors, error_class=Exception):
+        self.errors = errors
+        self.result = []
+        self.error_class = error_class
+        super().setUp()
+
+    def test_cleanup_in_happy_path(self):
+        self.run_with_errors()
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest",
+            "tearDown",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_error_in_set_up_class(self):
+        self.run_with_errors("setUpClass")
+        eq(self.result, [
+            "setUpClass Exception",
+            "cleanup"
+        ])
+
+    def test_cleanup_on_error_in_set_up(self):
+        self.run_with_errors("setUp")
+        eq(self.result, [
+            "setUpClass",
+            "setUp Exception",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_error_in_test(self):
+        self.run_with_errors("runTest")
+        eq(self.result, [
+            "setUpClass",
+            "setUp",
+            "runTest Exception",
+            "tearDown",
+            "tearDownClass",
+            "cleanup",
+        ])
+
+    def test_cleanup_on_test_fail(self):
+        self.run_with_errors("runTest", error_class=AssertionError)
```

Comment:
Why are we specifically testing on `AssertionError` here? It doesn't seem relevant to this test.

Reply:
Tests have two different failure modes: FAIL (test raised `AssertionError`) and ERROR (test raised any other error). I wanted to make sure the different code paths for FAIL vs ERROR handling did not affect class cleanup.